src/windows/webvieww.c

```diff
@@ -0,0 +1,387 @@
+
+	MTY_Free(message);
+
+	return 0;
+}
+
+static HRESULT STDMETHODCALLTYPE mty_webview_resource_requested(ICoreWebView2WebResourceRequestedEventHandler *this, ICoreWebView2 *sender, ICoreWebView2WebResourceRequestedEventArgs *args)
+{
+	struct webview *ctx = ((struct web_resource_requested *) this)->context;
+
+	ICoreWebView2WebResourceRequest *request = NULL;
+	ICoreWebView2WebResourceRequestedEventArgs_get_Request(args, &request);
+
+	wchar_t *url_w = NULL;
+	ICoreWebView2WebResourceRequest_get_Uri(request, &url_w);
+
+	if (strcmp("https://app.local/index.html", MTY_WideToMultiDL(url_w)))
+		return S_OK;
+
+	const wchar_t *headers = 
+		L"Content-Type: text/html\n"
+		L"Access-Control-Allow-Origin: *\n";
+
+	ICoreWebView2WebResourceResponse *response = NULL;
+	IStream *stream = SHCreateMemStream((const BYTE *) ctx->common.html, (UINT) strlen(ctx->common.html));
+	ICoreWebView2Environment_CreateWebResourceResponse(ctx->environment, stream, 200, L"OK", headers, &response);
+	ICoreWebView2WebResourceRequestedEventArgs_put_Response(args, response);
+
+	ICoreWebView2WebResourceResponse_Release(response);
+	IStream_Release(stream);
+
+	return S_OK;
+}
+
+static HRESULT STDMETHODCALLTYPE mty_webview_focus_changed(ICoreWebView2FocusChangedEventHandler *this, ICoreWebView2Controller *sender, IUnknown *args)
+{
+	struct webview *ctx = ((struct focus_changed *) this)->context;
+
+	ctx->has_focus = !ctx->has_focus;
+
+	return 0;
+}
+
+// Private
+
+static void *mty_webview_thread_func(void *opaque)
+{
+	struct webview *ctx = opaque;
+	ctx->thread_id = GetCurrentThreadId();
+
+	CoInitializeEx(NULL, COINIT_APARTMENTTHREADED);
+
+	CreateCoreWebView2EnvironmentWithOptions(NULL, L"C:\\Windows\\Temp\\webview-data", NULL, &ctx->environment_completed.handler);
+
+	MSG msg = {0};
+	while (GetMessage(&msg, NULL, 0, 0)) {
+		switch (msg.message) {
+			case WV_SIZE: {
+				RECT bounds = {0};
+
+				if (!ctx->common.hidden)
+					GetClientRect(ctx->hwnd, &bounds);
+
+				ICoreWebView2Controller3_put_BoundsMode(ctx->controller, COREWEBVIEW2_BOUNDS_MODE_USE_RAW_PIXELS);
+				ICoreWebView2Controller_put_Bounds(ctx->controller, bounds);
+				break;
+			}
+			case WV_SCRIPT_INIT:
+				ICoreWebView2_AddScriptToExecuteOnDocumentCreated(ctx->webview, (wchar_t *) msg.lParam, NULL);
+				MTY_Free((void *) msg.lParam);
+				break;
+			case WV_SCRIPT_EVAL:
+				ICoreWebView2_ExecuteScript(ctx->webview, (wchar_t *) msg.lParam, NULL);
+				MTY_Free((void *) msg.lParam);
+				break;
+			case WV_NAVIGATE:
+				ICoreWebView2_NavigateToString(ctx->webview, (wchar_t *) msg.lParam);
+				MTY_Free((void *) msg.lParam);
+				break;
+			default:
+				break;
+		}
+
+		TranslateMessage(&msg); // XXX: We MAY not need this, depending on whether webview requires 
+		                        // WM_CHAR messages or can handle WK_KEY* messages alone.
+		                        // Re-visit this later and remove this call if we can.
+		DispatchMessage(&msg);
+	}
+
+	return NULL;
+}
```

Comment:
```suggestion
static void *mty_webview_thread_func(void *opaque)
{
	struct webview *ctx = opaque;
	ctx->thread_id = GetCurrentThreadId();

	CoInitializeEx(NULL, COINIT_APARTMENTTHREADED);

	MSG msg = {0};
	bool first_run = true;
	bool running = true;

	while (running) {
		// Pump all pending messages and dispatch them
		while(PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
			if (LOWORD(msg.message) == WM_QUIT) {
				running = false;
				break;
			}

			switch (msg.message) {
				case WV_SIZE: {
					RECT bounds = {0};

					if (!ctx->common.hidden)
						GetClientRect(ctx->hwnd, &bounds);

					ICoreWebView2Controller3_put_BoundsMode(ctx->controller, COREWEBVIEW2_BOUNDS_MODE_USE_RAW_PIXELS);
					ICoreWebView2Controller_put_Bounds(ctx->controller, bounds);
					break;
				}
				case WV_SCRIPT_INIT:
					ICoreWebView2_AddScriptToExecuteOnDocumentCreated(ctx->webview, (wchar_t *) msg.lParam, NULL);
					MTY_Free((void *) msg.lParam);
					break;
				case WV_SCRIPT_EVAL:
					ICoreWebView2_ExecuteScript(ctx->webview, (wchar_t *) msg.lParam, NULL);
					MTY_Free((void *) msg.lParam);
					break;
				case WV_NAVIGATE:
					ICoreWebView2_NavigateToString(ctx->webview, (wchar_t *) msg.lParam);
					MTY_Free((void *) msg.lParam);
					break;
				default:
					break;
			}

			TranslateMessage(&msg); // XXX: We MAY not need this, depending on whether webview requires 
		                            // WM_CHAR messages or can handle WK_KEY* messages alone.
		                            // Re-visit this later and remove this call if we can.
			DispatchMessage(&msg);
		}

		if (first_run) {
			// This function must be called after the thread's message loop has been initialized
			// in order to avoid race conditions with processing messages sent from the WebView2 callbacks
			CreateCoreWebView2EnvironmentWithOptions(NULL, L"C:\\Windows\\Temp\\webview-data", NULL, &ctx->environment_completed.handler);
			first_run = false;
		}
		
		if (running)
			WaitMessage(); // blocks until a message is enqueued, effectively replicating the behaviour of GetMessage
	}

	return NULL;
}
```

This might be paranoia from me but it is theoretically possible.

Your code here assumes that when environment_completed.handler is called, the message loop has finished initializing the thread is waiting for messages.

But you don't have any guarantee that this is true. It all depends on how the microsoft library invokes the environment_completed.handler and later on, the controller_completed.handler. It is theoretically possible for line 238 to be called AFTER the WV library has invoked both the environment_completed.handler and the controller_completed.handler callback functions. If this happens, then all PostThreadMessage messages in those callback functions will get dropped and this would be a bug.

This may work 99% of the time and then fail randomly due to how timing dependent it is. It might also never show up on our machines but could show up on some random user's machine.

The safest thing to do is to make your callbacks wait until the message loop has begun spinning. My proposed change does this the easiest way I can think of.

Thoughts?